lib/auth/password_test.go

```diff
@@ -235,41 +235,101 @@ func TestServer_ChangePassword(t *testing.T) {
 	mfa := configureForMFA(t, srv)
 	username := mfa.User
 	password := mfa.Password
+	userClient, err := srv.NewClient(TestUser(username))
+	require.NoError(t, err)
+	passwordlessDev, err := RegisterTestDevice(
+		context.Background(),
+		userClient,
+		"passwordless-1",
+		proto.DeviceType_DEVICE_TYPE_WEBAUTHN,
+		mfa.TOTPDev,
+		WithPasswordless())
+	require.NoError(t, err)
 
 	tests := []struct {
-		name    string
-		newPass string
-		device  *TestDevice
+		name             string
+		oldPass          string
+		newPass          string
+		device           *TestDevice
+		challengeRequest *proto.CreateAuthenticateChallengeRequest
 	}{
 		{
 			name:    "OK TOTP-based change",
+			oldPass: password,
 			newPass: "llamasarecool11",
 			device:  mfa.TOTPDev,
+			challengeRequest: &proto.CreateAuthenticateChallengeRequest{
+				Request: &proto.CreateAuthenticateChallengeRequest_ContextUser{
+					ContextUser: &proto.ContextUser{},
+				},
+				ChallengeExtensions: &mfav1.ChallengeExtensions{
+					Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_CHANGE_PASSWORD,
+				},
+			},
 		},
 		{
-			name:    "OK Webauthn-based change",
-			newPass: "llamasarecool13",
+			name:    "OK TOTP-based change (legacy flow)",
+			oldPass: "llamasarecool11",
+			newPass: "llamasarecool12",
+			device:  mfa.TOTPDev,
+			challengeRequest: &proto.CreateAuthenticateChallengeRequest{
+				Request: &proto.CreateAuthenticateChallengeRequest_UserCredentials{
+					UserCredentials: &proto.UserCredentials{
+						Username: username,
+						Password: []byte("llamasarecool11"),
+					},
+				},
+			},
+		},
+		{
+			name:             "OK Webauthn-based change",
+			oldPass:          "llamasarecool12",
+			newPass:          "llamasarecool13",
+			device:           mfa.WebDev,
+			challengeRequest: &proto.CreateAuthenticateChallengeRequest{},
```

Comment:
Incomplete request?

Reply:
@codingllama Holy cow. Not only an incomplete request, but also incidentally, a piece of test that shouldn't have passed. I have noticed that we are actually falling here into a case where `loginFlow.finish()` uses a backward-compatible logic where it allows UNSPECIFIED scope, and this challenge request passes scope validation, even though it shouldn't. I'll definitely add a test that verifies that we can't do it and will provide a special case inside a special case there in `loginFlow.finish()` to treat the password change scope more strictly.

(As for backwards compatibility, I don't think we should care here, since only the web client offers password change ability anyway.)

Reply:
Gotcha, thanks for the careful testing here. This could cause problems if Proxy and Auth versions are mismatched, which is something to note, but that's likely OK given how serious password changes are.